docs/pallas/tpu/distributed.md

```diff
@@ -0,0 +1,1453 @@
+    num_scalar_prefetch=0,
+    in_specs=[
+        pl.BlockSpec(memory_space=pltpu.TPUMemorySpace.VMEM),
+    ],
+    out_specs=[
+        pl.BlockSpec(memory_space=pltpu.TPUMemorySpace.VMEM),
+        pl.BlockSpec(memory_space=pltpu.TPUMemorySpace.ANY),
+    ],
+    grid=(num_devices, 2),
+    scratch_shapes=(
+        [pltpu.SemaphoreType.DMA] * 5
+        + [pltpu.SemaphoreType.REGULAR] * 2  # Capacity semaphores
+        + [
+            pltpu.VMEM((block_size[0] // 2, block_size[1]), jnp.float32)
+        ]  # accum_scratch
+    ),
+)
+
+
+def pallas_reduce_scatter(input_arr):
+  input_arr = input_arr.reshape(num_devices, block_size[0], block_size[1])
+  return pl.pallas_call(
+      reduce_scatter_kernel,
+      out_shape=out_shape,
+      grid_spec=grid_spec,
+      compiler_params=dict(mosaic=dict(collective_id=0)),
+  )(input_arr)[0]
+
+
+pallas_result = jax.jit(
+    shard_map.shard_map(
+        pallas_reduce_scatter,
+        mesh=mesh,
+        in_specs=P(None, 'x'),
+        out_specs=P('x', None),
+        check_rep=False,
+    )
+)(input_arr)
+
+pallas_result = jax.block_until_ready(pallas_result)
+```
+
+```{code-cell} ipython3
+---
+executionInfo:
+  elapsed: 596
+  status: ok
+  timestamp: 1722904806442
+  user:
+    displayName: Justin Fu
+    userId: '17543197034567316452'
+  user_tz: 420
+id: E-NMh-_teoi4
+outputId: 24beb42f-1bdd-4c34-e8d2-681dd7f2e9c0
+---
+# Compare our result to XLA.
+def lax_reduce_sum_scatter(x):
+  x = x.reshape(num_devices, block_size[0], block_size[1])
+  return lax.psum_scatter(x, 'x')
+
+
+xla_result = jax.jit(
+    shard_map.shard_map(
+        lax_reduce_sum_scatter,
+        mesh=mesh,
+        in_specs=P(None, 'x'),
+        out_specs=P('x', None),
+    )
+)(input_arr)
+
+print('Input:', input_arr.shape, input_arr[::4, 0])
+print('Pallas Result:', pallas_result.shape, pallas_result[::4, 0])
+print('lax.psum_scatter Result:', xla_result.shape, xla_result[::4, 0])
+print(
+    'Difference |Pallas - lax.psum_scatter|:',
+    jnp.max(jnp.abs(pallas_result - xla_result)),
+)
+```
+
++++ {"id": "ThKas40r40Ji"}
+
+### Nested Remote and Local DMA Pipelines
+
+A limitation of the previous all-reduce and reduce-scatter kernels that we wrote is that the blocks we copy via remote DMA must be small enough to fit in our working VMEM that we use for accumulation. This limits the efficiency of our kernel since the amount of work that each device receives per iteration may not be enough for the operation to become compute-bound. For example, a matrix multiplication requires on the order of $O(N^3)$ compute operations, but only $O(N^2)$ memory transfers. Therefore, we want each block of work transferred between devices to be large enough such that the cost of compute becomes larger than the cost of communication so we can hide the communication using pipelining. For reference, the VMEM of a TPU (for generations v4/v5) is typically on the order of 10-100MB, whereas HBM ranges from 10-100GB.
```

Comment:
> This limits the efficiency of our kernel since the amount of work that each device receives per iteration may not be enough for the operation to become compute-bound.

> Therefore, we want each block of work transferred between devices to be large enough such that the cost of compute becomes larger than the cost of communication so we can hide the communication using pipelining.

This is confusing because unlike matmul the running examples (all-reduce and reduce-scatter) have compute and memory transfers that scale at the same rate with block size. So how can we make them compute-bound by varying the sizes? Do you mean "even though it doesn't apply to these examples, in some cases we may have compute that scales superlinearly with block size, like a matmul, so here's what you would do"? Or is there some reason that even for these reduction examples we may want to do this?

Reply:
Yes I'll add a note that this scaling property applies to matmul and not reduce-scatter. The reduce-scatter example is purely for educational purposes to demonstrate how to use the pipeline emitter.